components/brave_sync/brave_sync_service_unittest.cc

```diff
@@ -616,6 +617,93 @@ TEST_F(BraveSyncServiceTest, OnResetSync) {
       profile()->GetPrefs()->GetBoolean(syncer::prefs::kSyncBookmarks));
 }
 
+TEST_F(BraveSyncServiceTest, OnResetSyncWhenOffline) {
+  using brave_sync::jslib_const::kPreferences;
+  EXPECT_CALL(*sync_client(), OnSyncEnabledChanged).Times(AtLeast(1));
+  EXPECT_CALL(*observer(), OnSyncStateChanged(sync_service()))
+      .Times(AtLeast(3));
+  sync_service()->OnSetupSyncNewToSync("this_device");
+  EXPECT_TRUE(
+      profile()->GetPrefs()->GetBoolean(brave_sync::prefs::kSyncEnabled));
+  brave_sync_prefs()->SetThisDeviceId("0");
+
+  RecordsList records;
+  records.push_back(
+      SimpleDeviceRecord(SyncRecord::Action::A_CREATE, "0", "this_device"));
+  records.push_back(
+      SimpleDeviceRecord(SyncRecord::Action::A_CREATE, "1", "device1"));
+
+  sync_service()->OnResolvedPreferences(records);
+
+  auto devices = brave_sync_prefs()->GetSyncDevices();
+
+  EXPECT_TRUE(DevicesContains(devices.get(), "0", "this_device"));
+  EXPECT_TRUE(DevicesContains(devices.get(), "1", "device1"));
+
+  EXPECT_CALL(*sync_client(),
+              SendSyncRecords(kPreferences,
+                              ContainsDeviceRecord(SyncRecord::Action::A_DELETE,
+                                                   "this_device")))
+      .Times(1);
+
+  EXPECT_CALL(*sync_client(), SendFetchSyncRecords);
+  EXPECT_FALSE(sync_service()->pending_self_reset_);
+  sync_service()->OnResetSync();
+  EXPECT_TRUE(sync_service()->pending_self_reset_);
+
+  auto recordsSent = std::make_unique<RecordsList>();
+  sync_service()->OnRecordsSent(kPreferences, std::move(recordsSent));
+  EXPECT_FALSE(sync_service()->pending_self_reset_);
+
+  auto devices_final = brave_sync_prefs()->GetSyncDevices();
+  EXPECT_FALSE(DevicesContains(devices_final.get(), "0", "this_device"));
+  EXPECT_FALSE(DevicesContains(devices_final.get(), "1", "device1"));
+
+  EXPECT_TRUE(profile()
```

Comment:
These checks after reset should pull out as a function shared with the case `OnResetSync`, otherwise it is easily getting out of sync.

Reply:
addressed in https://github.com/brave/brave-core/compare/c2b6519c248584c5e5497da574ab8ea06846a25a..046dbae9503a4c53675d7abeb578cbb322ecae28